### DISCUSSION TITLE
Stress plot causes crash

### DISCUSSION THREAD

**Discussion by DougAJ4** [robbievanleeuwen/section-properties#187] [General]
Using pyxll to link Python to Excel, I have found all the sectionproperties plot routines cause a crash in Excel.

For the section and mesh plots I was able to avoid this by removing the code lines intended to display the image, and returning the Matplotlib fig object to the calling routine, where it can be displayed in Excel via pyxll.  

For the stress plot routine I haven't been able to get this to work.  Is there a way to call the plotting routines that will return the fig object, without trying to send the plot to the display?

**Comment by DougAJ4** [comment]
Update:
Reading the documentation, it seems that setting render = False should do what I want, but it doesn't seem to work.

The example I am running works up to the lines:
    case1 = section.calculate_stress(N = 1000, Mxx=5e6, Vx=10e3, Mzz=3e6)
    if out == 12:
        try:
            fig = case1.plot_stress_n_zz(render = False)

the final line crashes Excel without warning.

I also tried setting the default render value to False in plotting context, but it still crashes.

**Comment by Agent6-6-6** [comment]
@DougAJ4 - Maybe I'm reading the python code wrong (have not had time to test the final v2.0 yet), but doesn't the plot stress functions return the fig object anyway, i.e. you don't need the 'fig ='. All the render=False does is not display the matplotlib figure so you can do something else to the figure following this before potentially plotting it or saving it. 

Try save the figure or do something else with it afterwards to confirm (after removing the ```fig =```).

If your code works natively in python then I'm probably wrong here.

**Comment by DougAJ4** [comment]
Hi Agent666,
I should have tried plotting in Python first.
Running the code in example file 06_stress.py, everything seems to work until the line:
case1.plot_stress_m_zz(pause=False) 

With all default setting that line displays a white image briefly, then exits to the command line.

With render = False it exits without displaying anything.

This is Python 3.9.7 running under Windows 10.

**Comment by normanrichardson** [comment]
@DougAJ4 I'm not familiar with pyxll, but by the sounds of the original question you want the [`matplotlib.figure.Figure`](https://matplotlib.org/3.2.2/api/_as_gen/matplotlib.figure.Figure.html#matplotlib.figure.Figure). I believe at this point all sectionproperties plotting functions return a [`matplotlib.axes.Axes`](https://matplotlib.org/stable/api/axes_api.html#the-axes-class). The  `matplotlib.figure.Figure` can be retrieved from the `matplotlib.axes.Axes` via [`matplotlib.axes.Axes.get_figure`](https://matplotlib.org/3.2.2/api/_as_gen/matplotlib.axes.Axes.get_figure.html).

For example, modify 06_stress.py to:
```python
import sectionproperties.pre.library.steel_sections as steel_sections
from sectionproperties.analysis.section import Section
import matplotlib.pyplot as plt

# %%
# Create a 150x100x6 RHS on its side
geometry = steel_sections.rectangular_hollow_section(d=100, b=150, t=6, r_out=15, n_r=8)

# %%
# Create a mesh and section object. For the mesh, use a maximum area of 2
geometry.create_mesh(mesh_sizes=[2])
section = Section(geometry)

# %%
# Perform a geometry and warping analysis
section.calculate_geometric_properties()
section.calculate_warping_properties()

# %%
# Perform a stress analysis with Mx = 5 kN.m; Vx = 10 kN and Mzz = 3 kN.m
case1 = section.calculate_stress(Mxx=5e6, Vx=10e3, Mzz=3e6)

# %%
# Perform a stress analysis with My = 15 kN.m; Vy = 30 kN and Mzz = 1.5 kN.m
case2 = section.calculate_stress(Myy=15e6, Vy=30e3, Mzz=1.5e6)

# %%
# Plot the bending stress for case1
ax = case1.plot_stress_m_zz(render=False)
fig = ax.get_figure()
print(type(fig))
plt.show()

```
Terminal output:
```
<class 'matplotlib.figure.Figure'>
```

Also with `render=False` this allows modifications to be made to the plot before calling `plt.show()`. I believe `pause` is used to toggle on waiting for the close event from the plot window. I hope you can now pass fig to pyxll. I hope this helps.

**Reply by DougAJ4** [reply]
@normanrichardson Thanks, that was very helpful in getting the image displayed once I had got the crash problem fixed, which had nothing to do with Excel and pyxll.  

Running the example direct from Python was crashing Python at the line:
ax = case1.plot_stress_m_zz(render=False)
with any options for pause and or render.

I eventually got it working by:
installing mkl-service
upgrading:
numpy -> 1.22.1
Pillow -> 9.0.0
matplotlib -> 3.5.1

All done using pip install or pip install --upgrade

The upgrade packages were chosen with the aid of not very helpful debug messages, and web searches on the message.

I now have the plots working both direct from python and transferring to Excel (with the aid of fig = ax.get_figure()

**Reply by Spectre5** [reply]
Do you happen to know what versions of those libraries you had before?

**Reply by DougAJ4** [reply]
numpy 1.21.2
Pillow 8.4.0
matplotlib 3.4.3

and mkl-service installed:
ca-certificates                     2021.10.26-haa95532_2 --> 2021.10.26-haa95532_4
  certifi                          2021.10.8-py39haa95532_0 --> 2021.10.8-py39haa95532_2
  conda                               4.10.3-py39haa95532_0 --> 4.11.0-py39haa95532_0
  openssl                                 1.1.1l-h2bbff1b_0 --> 1.1.1m-h2bbff1b_0

HTH :)

**Comment by DougAJ4** [comment]
I have just edited the heading because it seems the original problem had nothing to do with transferring the plot to Excel.

**Reply by Spectre5** [reply]
What was the traceback when it crashed?  What was the error?

**Reply by DougAJ4** [reply]
I'm afraid I didn't record all the errors.  Will there be a log-file somewhere (using Visual Studio Code).

For the last one, before I upgraded matplotlib, the message was:
 DLL load failed while importing ft2font: The specified module could not be found.
